druid/src/lens.rs

```diff
@@ -141,3 +142,218 @@ where
         }
     }
 }
+
+/// Lens accessing a member of some type using accessor functions
+///
+/// See also the `lens` macro.
+///
+/// ```
+/// let lens = druid::lens::Field::new(|x: &Vec<u32>| &x[42], |x| &mut x[42]);
+/// ```
+pub struct Field<Get, GetMut> {
+    get: Get,
+    get_mut: GetMut,
+}
+
+impl<Get, GetMut> Field<Get, GetMut> {
+    /// Construct a lens from a pair of getter functions
+    pub fn new<T: ?Sized, U: ?Sized>(get: Get, get_mut: GetMut) -> Self
+    where
+        Get: Fn(&T) -> &U,
+        GetMut: Fn(&mut T) -> &mut U,
+    {
+        Self { get, get_mut }
+    }
+}
+
+impl<T, U, Get, GetMut> Lens<T, U> for Field<Get, GetMut>
+where
+    T: ?Sized,
+    U: ?Sized,
+    Get: Fn(&T) -> &U,
+    GetMut: Fn(&mut T) -> &mut U,
+{
+    fn with<V, F: FnOnce(&U) -> V>(&self, data: &T, f: F) -> V {
+        f((self.get)(data))
+    }
+
+    fn with_mut<V, F: FnOnce(&mut U) -> V>(&self, data: &mut T, f: F) -> V {
+        f((self.get_mut)(data))
+    }
+}
+
+/// Construct a lens accessing a type's field
+///
+/// This is a convenience macro for constructing `Field` lenses for fields or indexable elements.
+///
+/// ```
+/// struct Foo { x: u32 }
+/// let lens = druid::lens!(Foo, x);
+/// let lens = druid::lens!((u32, bool), 1);
+/// let lens = druid::lens!([u8], [4]);
+/// ```
+#[macro_export]
+macro_rules! lens {
+    ($ty:ty, [$index:expr]) => {
+        $crate::lens::Field::new::<$ty, _>(|x| &x[$index], |x| &mut x[$index])
+    };
+    ($ty:ty, $field:tt) => {
+        $crate::lens::Field::new::<$ty, _>(|x| &x.$field, |x| &mut x.$field)
+    };
+}
+
+/// `Lens` composed of two lenses joined together
+pub struct Then<T, U, B: ?Sized> {
+    left: T,
+    right: U,
+    _marker: PhantomData<B>,
+}
+
+impl<T, U, B: ?Sized> Then<T, U, B> {
+    pub fn new<A: ?Sized, C: ?Sized>(left: T, right: U) -> Self
```

Comment:
The docstring should probably point to the `then` method. Also, I'm not sure "left" and "right" are clear names, though I can't think of more authoritative ones off the top of my head. (I feel like I should read a primer on lenses or compositional mathematic operations)

I know in Haskell it's just function composition, but that doesn't seem available here.